packages/opentelemetry-metrics/src/export/Aggregator.ts

```diff
@@ -0,0 +1,87 @@
+/*!
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Distribution, Sum, LastValue } from './types';
+import { hrTime } from '@opentelemetry/core';
+import * as types from '@opentelemetry/types';
+
+/**
+ * Base interface for aggregators. Aggregators are responsible for holding
+ * aggregated values and taking a snapshot of these values upon export.
+ */
+export interface Aggregator {
+  /** Updates the current with the new value. */
+  update(value: number): void;
+
+  /** Returns snapshot of the current value. */
+  value(): Sum | LastValue | Distribution;
+}
+
+/** Basic aggregator which calculates a Sum from individual measurements. */
+export class CounterSumAggregator implements Aggregator {
+  private _current: number = 0;
+
+  update(value: number): void {
+    this._current += value;
+  }
+
+  value(): Sum {
+    return this._current;
+  }
+}
+
+/** Basic aggregator which keeps the last recorded value and timestamp. */
+export class GaugeAggregator implements Aggregator {
+  private _current: number = 0;
+  private _timestamp: types.HrTime = hrTime();
+
+  update(value: number): void {
+    this._current = value;
+    this._timestamp = hrTime();
+  }
+
+  value(): LastValue {
+    return {
+      value: this._current,
+      timestamp: this._timestamp,
+    };
+  }
+}
+
+/** Basic aggregator keeping all raw values (events, sum, max and min). */
+export class MeasureExactAggregator implements Aggregator {
+  private _distribution: Distribution;
+
+  constructor() {
+    this._distribution = {
+      min: Infinity,
```

Comment:
Would it be better to have these undefined?

Reply:
AFAIK `Math.min` or `Math.max` will return `NaN`, If at least one of arguments cannot be converted to a number. I think it is easier to work with `Infinity`.

Reply:
I agree that they're easier from a perspective of `Math.min` and `Math.max`, but if you export a measure before the first measurement is recorded (I am not even sure if this can happen?) then you may get weird results. I think at least we need to document on the Distribution type that `+-Infinity` is the uninitialized state, and make sure that in export pipeline we're shortcutting the `count=0` pathway.